### ISSUE TITLE
Re-use repository with different adapter.

### ISSUE THREAD

**Issue by danelowe** [lotus/model#188]
It seems as though the adapter is attached to the repository as a class-level attribute. 

I've got two applications mounted, using different adapters and mappings, but with the same repositories. 

```ruby
Lotus::Container.configure do
  mount Admin::Application, at: '/admin'
  mount Frontend::Application, at: '/' # Must go last
end
```

```ruby
#admin/application.rb
adapter type: :jsonapi, uri: 'http://localhost:9293/admin/'
mapping 'config/mapping'
```

```ruby
#frontend/application.rb
adapter type: :jsonapi, uri: 'http://localhost:9293/frontend/'
mapping 'config/mapping'
```

```ruby
require 'lotus/controller'

module Admin::Controllers::Jobs
  class Index
    include Lotus::Action

    expose :jobs

    def initialize(repository = Client::JobRepository)
      @repository = repository
    end

```

When I visit the jobs controller under `/admin` the adapter attached to and used by the  repository injected into the controller is the one with the uri  'http://localhost:9293/frontend/'

As a quick workaround, I've implemented the following method on the repositories

```ruby
   def for_app(namespace)
        const_set(namespace, Class.new(self)) unless const_defined?(namespace, false)
        const_get(namespace)
    end
```

then in my mappings I have: 

```ruby
collection :payments do
  entity      Client::Payment
  repository  Client::PaymentRepository.for_app('Admin')
```

And my admin controllers have to inject the sub-class that is created.

I don't think this is going to work long-term, however. 

Has there been any progress on the decision between class interface vs instance interface for repositories as mentioned here? 
https://github.com/lotus/model/pull/162

**Comment by danelowe** [comment]
or 

```ruby
def for_app(namespace)
        name = self.name.demodulize
        namespace.const_set(name, Class.new(self)) unless namespace.const_defined?(name, false)
        namespace.const_get(name)
      end
```

Do these have any relevance? Are these new methods already being used anywhere? 
https://github.com/lotus/model/pull/89
https://github.com/lotus/model/pull/122

**Comment by jodosha** [comment]
@danelowe Thanks for starting this discussion, but this is out of scope. A repository must be tight to a single adapter for several reasons:

  * Simplicity
  * Mutating class level values can cause thread safety problems
  * Internal implementation details are tight to the current adapter (see the example below)

Imagine the following query implemented by a repository.

```ruby
def self.by_author(author)
  query do
    where(author_id: author.id)
  end
end
```

We use `#where`, which is available in our SQL adapter. If I suddenly switch to an adapter that doesn't implement that query method, `.by_author` will blow up.

My suggestion for your case is to use two concrete repositories.

**Comment by danelowe** [comment]
Thanks @jodosha that makes sense. In essence, having the class interface is an easy way of interacting with repositories as singletons. 

Somewhere in the docs or in the code from lotusrb generator, I think I got the idea that repositories are intended to be stored in `/lib` and could be used between apps. I guess this practice could have some disastrous consequences. 

I also got a hint from the other pull requests mentioned above that maybe lotus could duplicate the repository classes into the app's namespace automatically on load?

**Comment by jodosha** [comment]
> Somewhere in the docs or in the code from lotusrb generator, I think I got the idea that repositories are intended to be stored in /lib and could be used between apps. I guess this practice could have some disastrous consequences.

It's correct. Main entities and repositories are stored under `lib/` and referenced from `apps/`. As long they use only one adapter, everything is fine. Eg. `BookRepository` uses SQL adapter. It lives in `lib/bookshelf/repositories/book_repository.rb` and can be used by `Web::Controllers::Books::Show`.

> I also got a hint from the other pull requests mentioned above that maybe lotus could duplicate the repository classes into the app's namespace automatically on load?

It can duplicate Lotus namespaces, if you need multiple usage of the same framework within the same Ruby process. An example is: `Lotus::Action`, is duplicated as `Web::Action` or `Admin::Action` for `Web` and `Admin` applications, respectively.

It's able to do the same thing for Lotus::Model. Attention, it duplicates `Lotus::Repository` into `Foo::Repository` (if your apps is called `Foo`), but not `BookRepository` itself.

In a nutshell, it duplicates Lotus modules, not application code.
